### ISSUE TITLE
Release crates in workspace to `crates.io`

### ISSUE THREAD

**Issue by luckysori** [itchysats/itchysats#2393]
This is useful as it would allow us to depend on several versions of the same protocol crate within our workspace (avoiding code duplication or branching logic) so that the maker can maintain backwards-compatibility with old takers.

Changes outside of this repository:

- [ ] Release `xtra` and `xtra-productivity` with all the latest changes.
- [ ] Release `maia` and `maia-core`.
- [ ] Release `maia-deprecated` (as an older version of `maia`) or just stop depending on it, breaking some clients.
- [ ] Depend on a released version of `electrum-client`. Blocked by https://github.com/bitcoindevkit/rust-electrum-client/issues/70. Alternatively we could go back to `bdk:0.18.0`, but that's probably too annoying.

This should enable us to release the following workspace crates:
- [ ] `bdk-ext`.
- [ ] `tokio-extras`.
- [ ] `xtra-libp2p`.
- [ ] `model`.
- [ ] `xtra-libp2p-rollover`, including version 1.0 and 2.0.
- [ ] `xtra-libp2p-offer`.
- [ ] `xtra-libp2p-ping`.

Concretely, this should allow us to remove the technical debt introduced with https://github.com/itchysats/itchysats/pull/2378/commits/35c4e4e658f8d72f23522a48fa6587cf12222ac5.

**Comment by bonomat** [comment]
Can we not achieve the same thing with versioning and git dependencies?

**Comment by luckysori** [comment]
> Can we not achieve the same thing with versioning and git dependencies?

I tried this whilst working on #2299, but in my experience it doesn't work. Consider this example:

- Tag `model` as `model-1`.
- Tag `rollover` as `rollover-1` (`rollover` explicitly depends on `model-1` and uses it in its API).
- Use `model-1` and `rollover-1` in `daemon`.

This produces an error because Rust is convinced that `rollover-1` actually depends on `model#rollover-1`. This happens even if the tags `model-1` and `rollover-1` are on separate commits. This could (hopefully) be resolved by moving crates out of the workspace, but that comes with other downsides.

**Comment by klochowicz** [comment]
how does the error look like? When I was trying to make `maia` and `maia-deprecated` to work, it was crucial to have exactly the same version of `maia-core` for both of them that was used, otherwise it would not have compiled. any chance it's a similar situation here

Whilst releasing a few of the crates above would be nice (the ones outside the workspace), I'm skeptical whether it's worth the effort (or that whether it will actually solve the problem).

**Comment by luckysori** [comment]
Assigning myself to turn this into a discussion, that should then become an ADR.

I still think my proposal could be good, but I should do a PoC first anyway.

**Comment by bonomat** [comment]
Can you explain why you want to release to crates.io please?

**Comment by luckysori** [comment]
To quote myself above:

> This is useful as it would allow us to depend on several versions of the same protocol crate within our workspace (avoiding code duplication or branching logic) so that the maker can maintain backwards-compatibility with old takers.

The idea was to, for example, be able to depend on `xtra-libp2p-offer:1.0.0` and `xtra-libp2p-offer:2.0.0` at the same time, without having heavily duplicated code (our current approach). Thinking about it now, I suppose the same thing could be achieved just by extracting the crates into separate repositories: we already leveraged the ability to depend on multiple versions of the same crate with `maia` in the past.

In any case, this proposal isn't all that practical because:

- A lot of the protocol crates depend on `model`, so we could end up running into situations where multiple versions of `model` would also need to be used. That is problematic because `model` is used everywhere in `daemon`. I can foresee the headaches already.
- We've already had to make changes to the `deprecated` versions of certain protocols, to fix incompatibilities and bugs on the maker side. This was easy because the code was still available in the workspace. With this proposal we would have to patch old versions in separate repositories, which makes the process of maintaining deprecated protocols a bit more cumbersome.

I've come around to the idea of just "duplicating" the code. One nice thing is that the taker side of the code for any given deprecated protocol can be readily deleted, because it's only the maker that has to keep running a deprecated protocol.

Even introducing version flags is acceptable and we ran with it in the past. I very much dislike that approach because logic can get very complicated, but there might be situations where we could/should use that approach.


With all this said, I'm closing the ticket.
